File: src/game/exploration.js
import { isExplored } from "./discovery.js";
import { freePosition, serviceSlots } from "./geometry.js";
import { routeCost } from "./navigation.js";
import { westBank } from "./map.js";
import { densityAt, densityReward } from "./density.js";
import { NEED_DECAY } from "./work-balance.js";
export function scoutLimit(w) { return Math.min(64, Math.max(1, Math.ceil(w.creatures.length / 12))); }
export function discoveryGain(w, p) {
  let gain = isExplored(w, p) ? 0 : 2;
  for (let i = 0; i < 8; i++)
    if (!isExplored(w, { x: p.x + Math.cos(i * Math.PI / 4) * 8,
      y: p.y + Math.sin(i * Math.PI / 4) * 8 })) gain++;
  return gain;
}
export function frontier(w, c, assignments) {
  if (Math.min(c.fed, c.clean, c.amused) < 68 || c.sickness > 20 || c.carry ||
      assignments.filter(a => a.task === "explore").length >= scoutLimit(w)) return null;
  const camps = w.objects.filter(o => ["orchard", "dwelling"].includes(o.type));
  const anchors = camps.length ? camps.slice().sort((a,b) =>
    Math.hypot(a.x-c.x,a.y-c.y)-Math.hypot(b.x-c.x,b.y-c.y)).slice(0,3) : [c];
  const phase = c.birthOrdinal * 2.399963, candidates = [];
  const scope = w.directives.members;
  // Follow a moving frontier. New food outposts extend the safe scouting range.
  for (const camp of anchors) for (let i = 0; i < 16; i++) {
    const angle = phase + i * Math.PI / 8;
    for (const radius of camps.length ? [16, 28, 40, 52] : [8, 14]) {
      const p = { x: camp.x + Math.cos(angle)*radius, y: camp.y + Math.sin(angle)*radius };
      if (Math.hypot(c.x-p.x,c.y-p.y)>60 || (!w.progress.bridge && !westBank(w,p))) continue;
      if (w.directives.region && (!scope.length || scope.includes(c.id)) && (p.x>42)!==(w.directives.region.x>42)) continue;
      if (assignments.some(a => a.task === "explore" && Math.hypot(a.point.x-p.x,a.point.y-p.y)<10)) continue;
      const gain = discoveryGain(w,p);
      if (gain) candidates.push({ ...p, score: gain*4-Math.hypot(c.x-p.x,c.y-p.y)*0.15+densityReward(densityAt(w,p).residents) });
    }
  }
  candidates.sort((a,b) => b.score-a.score);
  for (const candidate of candidates.slice(0,8)) {
    const p = freePosition(w,candidate,assignments.map(a=>a.point),2);
    if (!p || (!w.progress.bridge && !westBank(w,p)) || !discoveryGain(w,p)) continue;
    const cost = routeCost(w,c,p);
    // Budget the round trip to a reachable food entrance before hunger takes over.
    const home = camps.length ? Math.min(...anchors.flatMap(o=>serviceSlots(w,o).map(s=>routeCost(w,p,s)))) : cost;
    if (Number.isFinite(cost) && Number.isFinite(home) && (cost+home)/1.65*NEED_DECAY.fed<c.fed-45)
      return { x:p.x, y:p.y };
  }
  return null;
}

File: engine/src/exploration.rs
use crate::{Engine, density, development::minimum, value::*};
use serde_json::{Value, json};
pub fn scout_limit(w: &Value, policy: &str) -> usize {
    let policy = if flag(&w["directives"], "pauseWork") && policy == "expand" {
        "balanced"
    } else {
        policy
    };
    let rate = match policy {
        "expand" => 5.,
        "care" => 24.,
        _ => 12.,
    };
    ((list(w, "creatures").len() as f64 / rate).ceil() as usize).clamp(1, 64)
}
pub fn west_bank(w: &Value, p: Point) -> bool {
    p.x < crate::terrain::river_left(num(&w["map"], "seed") as u32, p.y)
}
impl Engine {
    pub fn discovery_gain(&self, p: Point) -> usize {
        let mut gain = if self.is_explored(p) { 0 } else { 2 };
        for i in 0..8 {
            let a = i as f64 * std::f64::consts::PI / 4.;
            if !self.is_explored(p.offset(a.cos() * 8., a.sin() * 8.)) {
                gain += 1;
            }
        }
        gain
    }
    pub fn frontier(&mut self, c: &Value, assignments: &[Value], policy: &str) -> Option<Value> {
        if minimum(c) < 68.
            || num(c, "sickness") > 20.
            || num(c, "carry") != 0.
            || assignments
                .iter()
                .filter(|a| text(a, "task") == "explore")
                .count()
                >= scout_limit(&self.world, policy)
        {
            return None;
        }
        let camps: Vec<_> = list(&self.world, "objects")
            .iter()
            .filter(|o| ["orchard", "dwelling"].contains(&text(o, "type")))
            .cloned()
            .collect();
        let origin = Point::read(c);
        let mut anchors = if camps.is_empty() {
            vec![c.clone()]
        } else {
            camps.clone()
        };
        anchors.sort_by(|a, b| {
            Point::read(a)
                .distance(origin)
                .total_cmp(&Point::read(b).distance(origin))
        });
        anchors.truncate(3);
        let phase = num(c, "birthOrdinal") * 2.399963;
        let mut candidates = Vec::new();
        for camp in &anchors {
            for i in 0..16 {
                let angle = phase + i as f64 * std::f64::consts::PI / 8.;
                let radii: &[f64] = if camps.is_empty() {
                    &[8., 14.]
                } else {
                    &[16., 28., 40., 52.]
                };
                for radius in radii {
                    let p = Point::read(camp).offset(angle.cos() * radius, angle.sin() * radius);
                    if origin.distance(p) > 60.
                        || (!flag(&self.world["progress"], "bridge") && !west_bank(&self.world, p))
                    {
                        continue;
                    }
                    let scope = list(&self.world["directives"], "members");
                    if self.world["directives"]["region"].is_object()
                        && (scope.is_empty() || scope.iter().any(|id| same_id(id, &c["id"])))
                        && (p.x > 42.) != (num(&self.world["directives"]["region"], "x") > 42.)
                    {
                        continue;
                    }
                    if assignments.iter().any(|a| {
                        text(a, "task") == "explore" && Point::read(&a["point"]).distance(p) < 10.
                    }) {
                        continue;
                    }
                    let gain = self.discovery_gain(p);
                    if gain > 0 {
                        let score = gain as f64 * 4. - origin.distance(p) * 0.15
                            + density::reward(self.resident_density(p), 6.);
                        candidates.push((p, score));
                    }
                }
            }
        }
        candidates.sort_by(|a, b| b.1.total_cmp(&a.1));
        let occupied: Vec<_> = assignments
            .iter()
            .map(|a| Point::read(&a["point"]))
            .collect();
        for (candidate, _) in candidates.into_iter().take(8) {
            let Some(p) = self.free_position(candidate, &occupied, 2.) else {
                continue;
            };
            if (!flag(&self.world["progress"], "bridge") && !west_bank(&self.world, p))
                || self.discovery_gain(p) == 0
            {
                continue;
            }
            let cost = self.route_cost(origin, p);
            let mut home = if camps.is_empty() {
                cost
            } else {
                f64::INFINITY
            };
            if !camps.is_empty() {
                for camp in &anchors {
                    for s in self.service_slots(camp, None) {
                        home = home.min(self.route_cost(p, Point::read(&s)));
                    }
                }
            }
            if cost.is_finite()
                && home.is_finite()
                && (cost + home) / 1.65 * 0.08 < num(c, "fed") - 45.
            {
                return Some(json!({"x":p.x,"y":p.y}));
            }
        }
        None
    }
}
